tests/test_utils/tests.py

```diff
@@ -1791,6 +1792,69 @@ def test_execute_recursive(self):
         self.assertEqual(len(callbacks), 2)
         self.assertIs(self.callback_called, True)
 
+    def test_execute_tree(self):
+        """
+        A visualisation of the callback tree tested. Each node is expected to be visited
+        only once. The child count of a node symbolises the amount of on commit
+        callbacks.
+
+        root
+        └── branch_1
+            ├── branch_2
+            │   ├── leaf_1
+            │   └── leaf_2
+            └── leaf_3
+        """
+        (
+            branch_1_call_counter,
+            branch_2_call_counter,
+            leaf_1_call_counter,
+            leaf_2_call_counter,
+            leaf_3_call_counter,
+        ) = [itertools.count(start=1) for _ in range(5)]
```

Comment:
`itertools` looks unnecessary :thinking: I would use `int` variables and alphabetized hooks, e.g.
```python
        branch_1_call_counter = 0
        branch_2_call_counter = 0
        leaf_1_call_counter = 0 
        leaf_2_call_counter = 0
        leaf_3_call_counter = 0
        
        def leaf_1():
            nonlocal leaf_1_call_counter 
            leaf_1_call_counter += 1
           
        def leaf_2():
            nonlocal leaf_2_call_counter
            leaf_2_call_counter += 1
        
        def leaf_3():
            nonlocal leaf_3_call_counter
            leaf_3_call_counter += 1
        
        def branch_1():
            nonlocal branch_1_call_counter
            branch_1_call_counter += 1
            transaction.on_commit(branch_2)
            transaction.on_commit(leaf_3)
            
        def branch_2():
            nonlocal branch_2_call_counter
            branch_2_call_counter += 1
            transaction.on_commit(leaf_1)
            transaction.on_commit(leaf_2)
```

Reply:
That's what I tried initially, although I'm getting smacked in the face with scope errors: 
```UnboundLocalError: local variable 'leaf_3_call_counter' referenced before assignment```

Reply:
Yes, `nonlocal` is necessary.